File: uhrwerk-config/src/main/java/io/qimia/uhrwerk/config/representation/Table.java
package io.qimia.uhrwerk.config.representation;


import io.qimia.uhrwerk.config.ConfigException;

public class Table{

    private String area;
    private String vertical;
    private String table;
    private String version;
    private Integer parallelism=1;
    private Integer max_bulk_size=1;
    private Partition partition;
    private Source[] sources;
    private Target[] targets;
    private Dependency[] dependencies;


    public Table() {}


    public String getArea() {
        return area;
    }

    public void setArea(String area) {
        this.area = area;
    }

    public String getVertical() {
        return vertical;
    }

    public void setVertical(String vertical) {
        this.vertical = vertical;
    }

    public String getTable() {
        return table;
    }

    public void setTable(String table) {
        this.table = table;
    }

    public String getVersion() {
        return version;
    }

    public void setVersion(String version) {
        this.version = version;
    }

    public Integer getParallelism() {
        return parallelism;
    }

    public void setParallelism(Integer parallelism) {
        this.parallelism = parallelism;
    }

    public Integer getMax_bulk_size() {
        return max_bulk_size;
    }

    public void setMax_bulk_size(Integer max_bulk_size) {
        this.max_bulk_size = max_bulk_size;
    }

    public Partition getPartition() {
        return partition;
    }

    public void setPartition(Partition partition) {
        this.partition = partition;
    }

    public Source[] getSources() {
        return sources;
    }

    public void setSources(Source[] sources) {
        this.sources = sources;
    }

    public boolean sourcesSet() {
        return sources != null;
    }

    public Target[] getTargets() {
        return targets;
    }

    public void setTargets(Target[] targets) {
        this.targets = targets;
    }

    public Dependency[] getDependencies() {
        return dependencies;
    }

    public void setDependencies(Dependency[] dependencies) {
        this.dependencies = dependencies;
    }

    public void validate(String path){
        path += "table/";
        if(area == null){
            throw new ConfigException("Missing field: " + path + "area");
        }
        if(vertical == null){
            throw new ConfigException("Missing field: " + path + "vertical");
        }
        if(table == null){
            throw new ConfigException("Missing field: " + path + "table");
        }
        if(version == null){
            throw new ConfigException("Missing field: " + path + "version");
        }
        if(parallelism == null){
            throw new ConfigException("Missing field: " + path + "parallelism");
        }
        if(max_bulk_size == null){
            throw new ConfigException("Missing field: " + path + "max_bulk_size");
        }
        if(partition == null){
            throw new ConfigException("Missing field: " + path + "partition");
        }
        else{
            partition.validate(path);
        }
        if(sources==null && dependencies==null){
            throw new ConfigException("Missing field: " + path + "sources or dependencies");
        }
        else {
            if(dependencies!=null){
                for(Dependency d: dependencies){
                    d.validate(path);
                }
            }
            if(sources!=null){
                for(Source s: sources){
                    s.validate(path);
                }
            }
        }
        if(targets==null){
            throw new ConfigException("Missing field: " + path + "targets");
        }
        else{
            for(Target t: targets){
                t.validate(path);
            }
        }
    }
}

File: uhrwerk-config/src/main/kotlin/io/qimia/uhrwerk/config/representation/Table.kt
package io.qimia.uhrwerk.config.representation

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import io.qimia.uhrwerk.config.builders.ConfigException

@JsonIgnoreProperties(ignoreUnknown = true)
data class Table(
    var area: String? = null,
    var vertical: String? = null,
    var table: String? = null,
    var version: String? = null,
    @JsonProperty("class_name")
    var className: String? = null,
    @JsonProperty("transform_sql_query")
    var transformSqlQuery: String? = null,
    @JsonProperty("partition_columns")
    var partitionColumns: Array<String>? = null,
    @JsonProperty("partitions")
    var partitionMappings: Array<PartitionMapping>? = null,
    @JsonProperty("variables")
    var tableVariables: Array<String>? = null,
    @JsonProperty("dynamic_partitioning")
    var dynamicPartitioning: Boolean = false,
    var parallelism: Int? = 1,
    @JsonProperty("max_bulk_size")
    var maxBulkSize: Int? = 1,
    var partition: Partition? = null,
    var sources: Array<Source>? = null,
    var targets: Array<Target>? = null,
    var dependencies: Array<Dependency>? = null
) {
    fun validate(path: String) {
        var path = path
        path += "table/"
        if (area == null) {
            throw ConfigException("Missing field: " + path + "area")
        }
        if (vertical == null) {
            throw ConfigException("Missing field: " + path + "vertical")
        }
        if (table == null) {
            throw ConfigException("Missing field: " + path + "table")
        }
        if (version == null) {
            throw ConfigException("Missing field: " + path + "version")
        }
        if (parallelism == null) {
            throw ConfigException("Missing field: " + path + "parallelism")
        }
        if (maxBulkSize == null) {
            throw ConfigException("Missing field: " + path + "max_bulk_size")
        }
        if (partition != null) {
            partition!!.validate(path)
        }
        if (sources == null && dependencies == null) {
            throw ConfigException("Missing field: " + path + "sources or dependencies")
        } else {
            if (dependencies != null) {
                for (d in dependencies!!) {
                    d.validate(path)
                }
            }
            if (sources != null) {
                for (s in sources!!) {
                    s.validate(path)
                }
            }
        }
        if (targets == null) {
            throw ConfigException("Missing field: " + path + "targets")
        } else {
            for (t in targets!!) {
                t.validate(path)
            }
        }

        if (partitionColumns != null && partitionMappings != null) {
            throw ConfigException("Only one of partition_columns or partitions can be set in $path")
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Table) return false

        if (area != other.area) return false
        if (vertical != other.vertical) return false
        if (table != other.table) return false
        if (version != other.version) return false
        if (className != other.className) return false
        if (transformSqlQuery != other.transformSqlQuery) return false
        if (parallelism != other.parallelism) return false
        if (maxBulkSize != other.maxBulkSize) return false
        if (partition != other.partition) return false
        if (sources != null) {
            if (other.sources == null) return false
            if (!sources.contentEquals(other.sources)) return false
        } else if (other.sources != null) return false
        if (targets != null) {
            if (other.targets == null) return false
            if (!targets.contentEquals(other.targets)) return false
        } else if (other.targets != null) return false
        if (dependencies != null) {
            if (other.dependencies == null) return false
            if (!dependencies.contentEquals(other.dependencies)) return false
        } else if (other.dependencies != null) return false

        return true
    }

    override fun hashCode(): Int {
        var result = area?.hashCode() ?: 0
        result = 31 * result + (vertical?.hashCode() ?: 0)
        result = 31 * result + (table?.hashCode() ?: 0)
        result = 31 * result + (version?.hashCode() ?: 0)
        result = 31 * result + (className?.hashCode() ?: 0)
        result = 31 * result + (transformSqlQuery?.hashCode() ?: 0)
        result = 31 * result + (parallelism ?: 0)
        result = 31 * result + (maxBulkSize ?: 0)
        result = 31 * result + (partition?.hashCode() ?: 0)
        result = 31 * result + (sources?.contentHashCode() ?: 0)
        result = 31 * result + (targets?.contentHashCode() ?: 0)
        result = 31 * result + (dependencies?.contentHashCode() ?: 0)
        return result
    }

    override fun toString(): String {
        return "Table(area=$area, vertical=$vertical, table=$table, version=$version, className=$className, transformSqlQuery=$transformSqlQuery, parallelism=$parallelism, maxBulkSize=$maxBulkSize, partition=$partition, sources=${sources?.contentToString()}, targets=${targets?.contentToString()}, dependencies=${dependencies?.contentToString()})"
    }

}